cns/restserver/util.go

```diff
@@ -350,11 +359,145 @@ func (service *HTTPRestService) removeToBeDeletedIPStateUntransacted(
+			getNetworkContainersResponse = append(getNetworkContainersResponse, getNetworkContainerResponse)
+			return getNetworkContainersResponse
+		}
+
+		logger.Printf("pod info %+v", podInfo)
+
+		exists = false
+		ncidBytes, err := json.Marshal(service.state.ContainerIDByOrchestratorContext[podInfo.Name()+podInfo.Namespace()])
+		if err != nil {
+			logger.Errorf("Failed to marlshal NC set")
+			return nil
+		}
+
+		networkContainerID = string(ncidBytes)
+		if networkContainerID != "" {
+			exists = true
+		}
+
+		ncSet := &ncSet{}
+		if err = json.Unmarshal(ncidBytes, &ncSet); err != nil {
+			logger.Errorf("Failed to unmarlshal NC set")
+			return nil
+		}
+
+		if exists {
+			for ncid := range *ncSet {
+				// If the goal state is available with CNS, check if the NC is pending VFP programming
+				waitingForUpdate, getNetworkContainerResponse.Response.ReturnCode, getNetworkContainerResponse.Response.Message = service.isNCWaitingForUpdate(service.state.ContainerStatus[networkContainerID].CreateNetworkContainerRequest.Version, networkContainerID) //nolint:lll // bad code
+				// If the return code is not success, return the error to the caller
+				if getNetworkContainerResponse.Response.ReturnCode == types.NetworkContainerVfpProgramPending {
+					logger.Errorf("[Azure-CNS] isNCWaitingForUpdate failed for NCID: %s with error: %s",
+						ncid, getNetworkContainerResponse.Response.Message)
+					getNetworkContainersResponse = append(getNetworkContainersResponse, getNetworkContainerResponse)
+					return getNetworkContainersResponse
+				}
+
+				vfpUpdateComplete := !waitingForUpdate
+				ncstatus := service.state.ContainerStatus[ncid]
+				// Update the container status if-
+				// 1. VfpUpdateCompleted successfully
+				// 2. VfpUpdateComplete changed to false
+				if (getNetworkContainerResponse.Response.ReturnCode == types.NetworkContainerVfpProgramComplete &&
+					vfpUpdateComplete && ncstatus.VfpUpdateComplete != vfpUpdateComplete) ||
+					(!vfpUpdateComplete && ncstatus.VfpUpdateComplete != vfpUpdateComplete) {
+					logger.Printf("[Azure-CNS] Setting VfpUpdateComplete to %t for NCID: %s", vfpUpdateComplete, ncid)
+					ncstatus.VfpUpdateComplete = vfpUpdateComplete
+					service.state.ContainerStatus[ncid] = ncstatus
+					if err = service.saveState(); err != nil {
+						logger.Errorf("Failed to save goal states due to %s", err)
+					}
+				}
+			}
+		} else if service.ChannelMode == cns.Managed {
+			// If the NC goal state doesn't exist in CNS running in managed mode, call DNC to retrieve the goal state
+			var (
+				dncEP     = service.GetOption(acn.OptPrivateEndpoint).(string)
+				infraVnet = service.GetOption(acn.OptInfrastructureNetworkID).(string)
+				nodeID    = service.GetOption(acn.OptNodeID).(string)
+			)
+
+			service.Unlock()
+			getNetworkContainerResponse.Response.ReturnCode, getNetworkContainerResponse.Response.Message = service.SyncNodeStatus(dncEP, infraVnet, nodeID, req.OrchestratorContext)
+			service.Lock()
```

Comment:
If you need to release a mutex like this and re-acquire it, there's usually other methods that should be added to better encapsulate the critical sections.

Reply:
Can we discuss more about this

Reply:
Yeah, so the whole reason to use a mutex is because you want some set of operations to be atomic from the perspective of other goroutines. You can usually give that set of things a name, and move all of that into its own method. Generally that method looks like:

```go
func (s *SomeType) doAtomicThings() {
       s.mu.Lock()
       defer s.mu.Unlock()

       // do stuff
}
```

That way, instead of dropping the lock mid-method here, you can have this method call into the atomic methods:

```go
s.doAtomicThings()
syncNodeStatus()
s.doOtherAtomicThings()
```

This turns "s" into something approximating a [Monitor](https://en.wikipedia.org/wiki/Monitor_(synchronization)) in the "thread-safe class" definition.

Reply:
As discussed, we can create a new task to optimize this later on since this does not block the PR